samples/java-spring-cinema-booking-choreography/README.md

```diff
@@ -0,0 +1,1439 @@
+To be able to create instances of `ReservationEntity`, a "Fold show events to reservation" [Kalix Action](https://docs.kalix.io/java/actions.html) component is required to facilitate `subscribe` to `ShowEntity` events and to send `Create` command to `ReservationEntity`.
+
+### Wallet
+
+![Wallet](images/KalixDesign-Wallet.png)
+
+For designing the Wallet bounded context in Kalix we are going to use for implementing Wallet (Entity/Aggregate) - [Kalix Entity (Event sourced)](https://docs.kalix.io/java/event-sourced-entities.html).
+
+**Note:** More information around implementing these Kalix service will be covered in following sections
+
+#### "Wallet" Entity
+
+As the Show entity, the Wallet entity will be implemented using a [Kalix Entity (Event sourced)](https://docs.kalix.io/java/event-sourced-entities.html). More info about the event sourced state model can be found [here](https://docs.kalix.io/concepts/state-model.html#_the_event_sourced_state_model).
+
+### Context map
+
+![Context Map](images/KalixDesign-ContextMap.png)
+
+For the interaction between Show and Wallet, the implementation of our "Book & Pay the show" _choreography_-based Saga pattern uses a [Kalix Action](https://docs.kalix.io/java/actions.html) to facilitate asynchronous event-driven communication.
+
+**Note:** Kalix does not require the use of message broker for asynchronous event-driven communication.
+
+# Kickstart a Kalix development project
+
+## Maven archetype
+
+A Kalix Maven archetype is used to generate a recommended project structure:
+
+```shell
+mvn archetype:generate \
+  -DarchetypeGroupId=io.kalix \
+  -DarchetypeArtifactId=kalix-spring-boot-archetype \
+  -DarchetypeVersion=latest
+```
+
+```
+Define value for property 'groupId': `com.example`<br>
+Define value for property 'artifactId': `java-spring-cinema-booking-choreography`<br>
+Define value for property 'version' 1.0-SNAPSHOT: :<br>
+Define value for property 'package' com.example: : `com.example`<br>
+```
+
+## Import the generated project into your IDE/editor
+
+The generated Maven project has all required Maven plugins and dependencies needed to develop, test, run and deploy Kalix services.
+
+## Adding additional, non-mandatory dependencies
+
+Additional dependencies that are NOT mandatory for developing Kalix service but are convenient to be used in different situations (more explained in following sections).<br>
+Add following dependencies in pom.xml:
+```
+<dependency>
+  <groupId>io.vavr</groupId>
+  <artifactId>vavr</artifactId>
+  <version>0.10.4</version>
+</dependency>
+<dependency>
+  <groupId>org.assertj</groupId>
+  <artifactId>assertj-core</artifactId>
+  <version>3.23.1</version>
+  <scope>test</scope>
+</dependency>
+<dependency>
+  <groupId>org.awaitility</groupId>
+  <artifactId>awaitility</artifactId>
+  <version>4.2.0</version>
+</dependency>
+```
+
+`vavr` library is used to provide support for:
+- immutable collections
+- Either support 
```

Comment:
I've been considering whether vavr is a good fit for our use.

It seems suitable for the intermediate to advanced samples.

However, for the introductory level ones targeted at new Kalix users unfamiliar with vavr, it might be a distraction. Learning another new thing could divert their attention from the core Kalix concepts.

What do others think?

Reply:
I had it in the user register choreography, but removed because it was not adding much value. Here is different. Logic is inside the domain, and the domain can reject a command by returning a Left with a message. 

The entity wrapper uses it to build a response and return an error with the message in passed with Left. 

But we could simplify it as well by having a more straightforward (and explicit) type like: `Result` that either has a list of events or an error message. Left / Right is not really expressive. 

By the way, we do have a similar a class already. It's called `Response`. The one that I suggested to move to `com.example.common`. 

That said, if we are adding vavr just because of Either, then we can simply build our own more expressive variation of it.

My 2cc.

Reply:
we use `vavr` also for immutable map

Reply:
Correct. And that's an important topic (and old topic). Mutable objects (including collections) is a problem. Using record is not enough because a record can reference mutable objects themselves.

If someone uses a mutable data structure and mutates it inside the command handler, the state will be effectively mutated and eventually snapshotted without the user being aware of it.

Reply:
[Today I learned](https://typesafe.slack.com/archives/C02A6LVUZ/p1710934023470989) the future of the Vavr project appears to be uncertain https://danieldietrich.dev/blog/2023/12/13/drawing-a-clear-line/.

Due to this uncertainty, we should consider removing its usage from this sample.